### PR TITLE
pkg/config/remote: Handles multiple APM Remote Config files

### PR DESCRIPTION
**Pull request by Barbayar**
### What does this PR do?
It allows APM to send multiple config files to the agent.

Screenshot 2022-02-22 at 00 35 57

### Motivation
APM remote configs can be generated from multiple sources, and each source has its own config id. They can enabled/disabled independently by their config id.

### Additional Notes
- [ ] TODO: write test cases with multiple files.

### Reviewer's Checklist

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] The appropriate `team/..` label has been applied, if known.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

### CODE REVIEW

File: pkg/config/remote/configs_apm_sampling.go [hunk 1]
```diff
@@ -54,3 +52,15 @@ func (c *apmSamplingConfigs) update(configFiles map[string]configFiles) (*APMSam
 	}
 	return nil, nil
 }
+
+func (c *apmSamplingConfigs) shouldUpdate(configFiles map[string]configFiles) bool {
+	if c.config == nil || len(c.config.Configs) != len(configFiles) {
```
Comment (Barbayar) [inline pkg/config/remote/configs_apm_sampling.go:57]: Update cases:

1. No remote configs in the agent
2. New config id is added
3. An existing config id is removed
4. One of the config ids has a newer version

**Comment by Barbayar** [comment]
@raphaelgavache 
Which version should we keep here when we have multiple config files?
https://github.com/DataDog/datadog-agent/blob/6f0daca4b75519e2c5511e2e03e90dcd4a8f04fc/pkg/trace/sampler/remote_rates.go#L191

File: pkg/config/remote/configs_apm_sampling.go [hunk 2]
```diff
@@ -26,20 +26,18 @@ func newApmSamplingConfigs() *apmSamplingConfigs {
...
-				Config: Config{
-					ID:      configID,
-					Version: files.version(),
-				},
-			},
+		update.Config.Configs[configID] = Config{
+			ID:      configID,
+			Version: files.version(),
 		}
 		for _, file := range files {
```
Comment (Barbayar) [inline pkg/config/remote/configs_apm_sampling.go:42]: This is a bit confusing. Do we have a set of files under one config id?

Comment (arbll) [reply pkg/config/remote/configs_apm_sampling.go:42]: Yes I know it's unexpected, some products wanted to bake in "diffs" which led to this mess. If you don't want that capability feel free to error out if there are more than one file per config.

File: pkg/trace/sampler/remote_rates.go [hunk 3]
```diff
@@ -67,7 +69,19 @@ func newRemoteRates(maxTPS float64) *RemoteRates {
 }
 
 func (r *RemoteRates) onUpdate(update remote.APMSamplingUpdate) error {
-	log.Debugf("fetched config version %d from remote config management", update.Config.Version)
+	var sb strings.Builder
+	addComma := false
+	for _, c := range update.Config.Configs {
```
Comment (Barbayar) [inline pkg/trace/sampler/remote_rates.go:74]: We could use `MetaStruct` instead of this.

File: pkg/config/remote/configs_apm_sampling.go (hunk 1 above)
Comment (raphaelgavache) [reply pkg/config/remote/configs_apm_sampling.go:56]: What happens when shouldUpdate is false? Does the empty configs make it to pkg/trace/remote_sampler and erase existing configs?

**Review by raphaelgavache — APPROVED** [review]

**Review by arbll — APPROVED** [review]

**Pull request merged**